File: oops/solid/ts/L.ts
/*
3. Liskov Substitution Principle (LSP):
Definition: Objects of a superclass should be replaceable with objects of a subclass without affecting the correctness of the program.

Explanation: Subclasses should adhere to the behavior expected by the superclass. If a subclass violates this, it can lead to unexpected behavior.
*/

class Bird {

    isBird() {
        return true
    }

    fly() {
        return "bird can fly"
    }

}

class Duck extends Bird {

    constructor(readonly name: string) {
        super()
    }

    swim() {
        return "this duck can swim"
    }
}

function checkBird(bird: Bird) {
    return bird.isBird()
}

const d = new Duck("Donald")
const bird = new Bird()

checkBird(bird)
checkBird(d)

File: oops/solid/py/L.py
from abc import ABC

# 3. Liskov Substitution Principle (LSP):


class Bird(ABC):

    def isBird(self):
        return True


def checkBird(b: Bird):
    return b.isBird()


# violation of 3
class Penguin(Bird):

    # this should violate principle 3
    def isBird(self):
        raise NotImplementedError(
            "Penguins are birds, but this method is not implemented."
        )


# example
checkBird(Penguin())  # will throw error

# ---------------------------- correct implementation ------------------------------


class Penguin2(Bird):

    #    avoid the implementation of isBird function if not necessary

    # or
    def isBird(self):
        return False

    def swim(self):
        return True


checkBird(Penguin2())  # no issue
